docs/src/extend/custom-rule-plugin-tutorial.md

```diff
@@ -0,0 +1,431 @@
+
+The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule doesn't match any of the criteria in the tests, the method throws an error.
+
+```javascript
+// foo-bar.test.js
+const {RuleTester} = require("eslint");
+const fooBarRule = require("./foo-bar");
+
+const ruleTester = new RuleTester({
+  // Must use at least ecmaVersion 2015 because
+  // that's when `const` variable introduced.
+  parserOptions: { ecmaVersion: 2015 }
+});
+
+// Throws error if the tests in ruleTester.run() do not pass
+ruleTester.run(
+  "foo-bar", // rule name
+  fooBarRule, // rule code
+  { // checks
+    // 'valid' checks cases that should pass
+    valid: [{
+      code: "const foo = 'bar';",
+    }],
+    // 'invalid' checks cases that should not pass
+    invalid: [{
+      code:"const foo = 'baz';",
+      output: 'const foo = "bar";',
+      errors: 1,
+    }],
+  }
+);
+
+console.log("All tests passed!");
+```
+
+## Step 7: Bundle the Custom Rule in a Plugin
+
+Now that you've written the custom rule and validated that it works, you can include it in a plugin. Using a plugin, you can share the rule in an npm package to use in other projects.
+
+Create the file for the plugin:
+
+```shell
+touch eslint-plugin-foo-bar.js
+```
+
+And now write the plugin code. Plugins are just exported JavaScript objects. To include a rule in a plugin, include it in the plugin's `rules` object, which contains key-value pairs of rule names and their source code.
+
+To learn more about creating plugins, refer to [Create Plugins](plugins).
+
+```javascript
+// eslint-plugin-foo-bar.js
+
+const fooBarRule = require("./foo-bar");
+module.exports = { rules: { "foo-bar": fooBarRule } };
+```
+
+## Step 8: Publish the Plugin
+
+To publish a plugin containing a rule to npm, you need to do a bit of set up in your project's `package.json` file.
+
+To help publish the package easily, install the dependency [np](https://www.npmjs.com/package/np):
+
+```shell
+npm install --save-dev np
+```
+
+Add a script to run `np`:
+
+```javascript
+// package.json
+// ...rest of package.json
+  "scripts": {
+    "test": "node foo-bar.test.js",
+    "pub": "np --yarn false" // new publication script
+  },
+// ...rest of package.json
+```
+
+Next, configure the `package.json` to be ready for publishing the plugin. Add the following in the corresponding fields:
+
+1. `"name"`: A unique name for the package. No other package on npm can have the same name.
+1. `"main"`: The relative path to the plugin file. Following this example, the path is `"eslint-plugin-foo-bar.js"`.
+1. `"description"`: A description of the package that's viewable on npm.
+1. `"peerDependencies"`: Add `"eslint": ">=0.8.0"` as a peer dependency. Plugins were introduced in eslint@0.8.0, so any version greater than or equal to that is necessary to use the plugin. By declaring `eslint` as a peer dependency, it requires that users add the package to the project separately from the plugin.
```

Comment:
We may want to use `>=8.0.0` here. It's unlikely newer rules will work all the way back to 0.8.0.

Reply:
note to self: update in src code too